lib/core/include/scipp/core/spatial_transforms.h

```diff
@@ -0,0 +1,194 @@
+// SPDX-License-Identifier: BSD-3-Clause
+// Copyright (c) 2021 Scipp contributors (https://github.com/scipp)
+/// @file
+/// @author Simon Heybrock
+#pragma once
+#include <Eigen/Core>
+#include <Eigen/Geometry>
+
+#include "scipp/core/dtype.h"
+
+namespace scipp::core {
+
+class RotationTransform {
+private:
+  // Store as quaterniond as this is more space efficient than storing as matrix
+  // (4 doubles for quat vs 9 doubles for 3x3 matrix).
+  Eigen::Quaterniond quat;
+
+public:
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+  RotationTransform() : quat(Eigen::Quaterniond::Identity()){};
+  explicit RotationTransform(const Eigen::Quaterniond &x) : quat(x){};
+
+  [[nodiscard]] Eigen::Matrix3d matrix() const {
+    return quat.toRotationMatrix();
+  }
+
+  inline bool operator==(const RotationTransform &other) const {
+    return this->matrix() == other.matrix();
+  }
+};
+
+class ScalingTransform {
+private:
+  Eigen::DiagonalMatrix<double, 3> mat;
+
+public:
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+  ScalingTransform()
+      : mat(Eigen::Matrix3d::Identity().diagonal().asDiagonal()){};
+  explicit ScalingTransform(const Eigen::DiagonalMatrix<double, 3> &x)
+      : mat(x){};
+
+  [[nodiscard]] Eigen::Matrix3d matrix() const { return mat; }
+
+  inline bool operator==(const ScalingTransform &other) const {
+    return this->matrix() == other.matrix();
+  }
+};
+
+class TranslationTransform {
+private:
+  Eigen::Vector3d vec;
+
+public:
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+  TranslationTransform() : vec(Eigen::Vector3d(0, 0, 0)){};
+  explicit TranslationTransform(const Eigen::Vector3d &x) : vec(x){};
+
+  [[nodiscard]] Eigen::Vector3d vector() const { return vec; }
+
+  inline bool operator==(const TranslationTransform &other) const {
+    return this->vector() == other.vector();
+  }
+};
+
+[[nodiscard]] inline RotationTransform operator*(const RotationTransform &lhs,
+                                                 const RotationTransform &rhs) {
+  return RotationTransform(Eigen::Quaterniond(lhs.matrix() * rhs.matrix()));
+};
+
+[[nodiscard]] inline Eigen::Matrix3d operator*(const RotationTransform &lhs,
+                                               const ScalingTransform &rhs) {
+  return lhs.matrix() * rhs.matrix();
```

Comment:
Can't you let `Eigen` figure out the details, instead of working with matrices manually?
```suggestion
                                               const ScalingTransform &rhs) {
  return lhs.get() * rhs.get();
```
where the `get()` method returns the underlying `Eigen` object.